### PR TITLE
OKRs - 2019 Q1 JS Core WG

### PR DESCRIPTION
**Pull request by alanshaw**
refs #792

- [x] [Score 2018 Q4 OKRs](https://docs.google.com/spreadsheets/d/139lROP7-Ee4M4S7A_IO4iIgSgugYm7dct620LYnalII/edit#gid=274358435)
- [x] [Async Retrospective](https://docs.google.com/document/d/1z6_eTW_8pI5VUj5UMq4K_Gpofofd7ENwzLdEBCuNULI/edit)
- [x] 2019 Q1 OKRs Open Planning (this thread)
- [x] Move 2019 Q1 OKRs to [2019 Q1 OKRs Spreadsheet](https://docs.google.com/spreadsheets/d/1BtOfd7s9oYO5iKsIorCpsm4QuQoIsoZzSz7GItE-9ys/edit#gid=274358435)

cc @hugomrdias @achingbrain @daviddias

### CODE REVIEW

**Comment by vasco-santos** [comment]
## IPNS has feature parity with go

I would like to do some improvements in IPNS in the next quarter.

At first, DDC WG would really benefit from having a streaming api for IPNS, as `go-ipfs` recently added. Context from @dirkmc comment: [https://github.com/ipfs/js-ipfs/pull/1725#issuecomment-440311875](https://github.com/ipfs/js-ipfs/pull/1725#issuecomment-440311875)

I also added an OKR in `libp2p` land for having the stream available in the DHT, which is a requirement for implementing this here. 

`go-ipfs` also had some recent improvements, such as some stuff in the configuration file and experimental parameters that we still lack in JS land. 

## Service worker gateway v2

We have been discussing the next version of the `service-worker-gateway` for a long time. Maybe someone can have a hand on this during Q1.

References:

https://github.com/ipfs-shipyard/service-worker-gateway/issues/30
https://github.com/ipfs/developer-meetings/blob/master/DEEP_DIVES/service-worker.md

**Comment by daviddias** [comment]
@lidel I believe that the In Web Browsers WG is taking over the Service Worker Gateway project, is that correct? Possibly ipfsd-ctl too?

**Comment by alanshaw** [comment]
* 13 weeks in the quarter: 65 days (assuming 5 day work week)
* Planned holidays: 54 days
* Probable hack week/conference: 49 days
* I aim to spend 90% of that time in JS IPFS WG: 44 days
* 60% of that is **~26 days** or **~207 hours**

**Comment by achingbrain** [comment]
## Carried over from Q4 2018

- [ ] A repo migration tool exists and JS IPFS can automatically upgrade older repos
  - ipfs/js-ipfs#1115
  - Supports the great CIDv0 -> CIDv1 migration
- [ ] Continuous deployment for js projects
  - Dust needs to settle on which CI tool we are using
  - Resolve how we can store deploy key type secrets without them being vulnerable to exfiltration
  - Start with `js-ipfs-unixfs-*` and `js-ipfs-mfs`
  - Expand to `js-ipld-*` and `js-libp2p-*`

## New for Q1 2019

- [ ] Add IPFS as a transport to [npm/tink](https://github.com/npm/tink)
- [ ] Put [Homebrew](https://brew.sh/) on IPFS

**Comment by alanshaw** [comment]
## Things to finish

* Base32 CIDv1 migration is complete

This is basically 2/3 complete. We have the CID tool, CID version agnostic get and CID base option. The final step is to take the plunge and make the switch to CIDv1 and base32 encoding by default.

For additional bonus points we could migrate the repo to key on multihashes instead of CIDs. However this might depend on there being a tool to migrate a repo (or we could do it ad hoc on access) and has consequences for pinning that would need to be worked out.

* A JS IPFS daemon is one (or more!) of the IPFS gateway nodes

I'm not super excited to work on this but I feel as though this is really necessary towards being production ready.

* A website is published to visually track a set of basic performance profiles of JS IPFS releases over historical versions

This basically already exists, but needs to be hooked up to CI and more benchmarks added before we can really announce it.

* Test quality is improved, duplicates removed and coverage increased by at least 5%

I really just need to have the time to get stuck into this...

## New things I want to tackle

* Drive async iterators endeavour - aim for proposal/PR to at least half of all repos in scope

* Pubsub in the browser - dependency on switch to fetch API, enable cancelable requests

* Fix up libdweb discovery and transport modules - include by default in companion

**Comment by lidel** [comment]
> @lidel I believe that the In Web Browsers WG is taking over the Service Worker Gateway project, is that correct? Possibly ipfsd-ctl too?

@daviddias Depends on resources available to WB WG in Q1. If it comes with time commitment from someone, then I'd be happy to add KR related to Service Worker to Web Browsers WG https://github.com/ipfs/team-mgmt/pull/804

**Comment by alanshaw** [comment]
1. @achingbrain please do Q4 OKR scoring

1. @hugomrdias please do async retro

1. @achingbrain @vasco-santos I've taken a stab at the Q1 OKRs - I could do with some suggestions for grouping/objective naming...

**Comment by alanshaw** [comment]
These OKRs have a big ask from infra 🙏 - we'd like a JS IPFS node to be added as one of the nodes that make up the IPFS gateway cluster.

I envisage this to be a process of slow introduction:

1. Initially I want to get JS IPFS running on a macine or vm provisioned in a similar way to the existing nodes, in order for it to be easy to add as a gateway node when the time comes. I suspect this will require some pointers to existing docs or access to a repo where the provisioning scripts are kept so that I can learn how it fits together and replicate it for a JS IPFS node.

1. Run the machine for a suitable period to assess it's stability. Test it's ability to process requests and handle load. I'll need some information from infra about the load that it would be expected to handle so that a suitable stress test can be devised.

1. Use the machine or set or machines as the default gateway in a tool like companion (@lidel will this be possible? - perhaps in the beta channel?) to further test in real world scenarios.

1. Slow introduction to the gateway cluster - can we introduce it and slowly ramp up the load as it proves it's worth?

cc @eefahy @ipfs/infra

File: OKR/JS_CORE.md [hunk 1]
```diff
@@ -2,6 +2,31 @@
 
 We frame our ongoing work using a process based on quarterly Objectives and Key Results (OKRs). Objectives reflect outcomes that are challenging, but realistic. Results are tangible and measurable.
 
+## 2019 Q1
+
+### TODO NAME ~implementation
+
+* p0 - Base32 encoded version 1 CIDs are the default for new content added to IPFS @alanshaw
+* p0 - IPNS has a streaming API @vasco-santos
+* p1 - IPNS supports publishing parameters `ttl`, `dhtt` and `dhtrc` @vasco-santos
```
Comment (vasco-santos) [inline OKR/JS_CORE.md]: As the `dht` parameters are for resolve, I would put it in the following way:

```suggestion
* p1 - IPNS supports `ttl`, `dhtt` and `dhtrc` parameters @vasco-santos
```

**Comment by alanshaw** [comment]
Ok, I've reworded https://github.com/ipfs/team-mgmt/blob/7bdd947db5cf95a95fce0745891c8ff2178d8091/OKR/JS_CORE.md

@vasco-santos @achingbrain @daviddias can I get a review please 🙏

**Comment by alanshaw** [comment]
> A **more** complete implementation of IPFS in JavaScript

??? I'm open to suggestions - also what is the "etc."?

**Comment by alanshaw** [comment]
Ok so in terms of the HTTP API, JS IPFS does not implement the following endpoints:

* `/bitswap/ledger`
* `/bitswap/reprovide`
* `/bootstrap/add/default`
* `/bootstrap/list`
* `/bootstrap/rm/all`
* `/commands`
* `/config/profile/apply`
* `/dag/put`
* `/dag/resolve`
* `/dht/*`
* `/diag/*`
* `/files/chcid`
* `/filestore/*`
* `/log/*`
* `/mount`
* `/object/diff`
* `/p2p/*`
* `/refs/*`
* `/repo/fsck`
* `/repo/gc`
* `/repo/verify`
* `/swarm/filters/*`
* `/tar/*`
* `/update`

That would be a LOT to do in the quarter! There's a subset of that API that I'd consider "good enough" for 99% of people using JS IPFS to call it "complete". I'd say the most important ones to implement might be `/dag/put`, `/dag/resolve`, `/repo/gc`, `/refs` and `/swarm/filters/*` (`/dht` is basically done).

`/dag*` should be relatively easy to replicate what Go IPFS is doing, `/repo/gc` will be the big one, `/refs` should be easy and `/swarm/filters/*` might be tricky and a foray into libp2p land :D

Note: I've listed the HTTP API here but it makes sense to also implement the CLI for these at the same time.

@achingbrain do you want to drop the homebrew OKR in favour of implementing those (or a different set)?

Either way, we're not going to be **fully complete** this quarter but as an aspirational objective do you think it works? You never know we might have time to implement _everything_ 😉.

**Comment by achingbrain** [comment]
😨 There's a lot to do!

Yes, lower priority items in favour of fixing up `js-ipfs` sounds sensible.

File: OKR/JS_CORE.md [hunk 2]
```diff
@@ -2,6 +2,27 @@
...
+
+### Used in production
+
+* p0 - A JS IPFS daemon is one (or more!) of the IPFS gateway nodes [@alanshaw](https://github.com/alanshaw)
+* p2 - Perf benchmarking site is launched and 10 different benchmarking scenarios exist [@alanshaw](https://github.com/alanshaw)
+* p0 - IPFS is a transport for [npm/tink](https://github.com/npm/tink) [@achingbrain](https://github.com/achingbrain)
+
+### Inviting for developers and contributors
+
+* p1 - Continuous deployment drives `js-ipfs-unixfs-*` and `js-ipfs-mfs` [@achingbrain](https://github.com/achingbrain)
```
Comment (daviddias) [inline OKR/JS_CORE.md]: What work does this KR translate into?

File: OKR/JS_CORE.md [hunk 3]
```diff
@@ -2,6 +2,27 @@
...
+* p1 - IPNS supports `ttl`, `dhtt` and `dhtrc` parameters and republish configurations [@vasco-santos](https://github.com/vasco-santos)
+* p1 - A repo migration tool exists and JS IPFS can automatically upgrade older repos [@achingbrain](https://github.com/achingbrain)
+* p2 - Vital missing APIs are implemented i.e. `/dag/put`, `/dag/resolve`, `/repo/gc`, `/refs` and `/swarm/filters/*` [@achingbrain](https://github.com/achingbrain)
+
+### Used in production
+
+* p0 - A JS IPFS daemon is one (or more!) of the IPFS gateway nodes [@alanshaw](https://github.com/alanshaw)
+* p2 - Perf benchmarking site is launched and 10 different benchmarking scenarios exist [@alanshaw](https://github.com/alanshaw)
+* p0 - IPFS is a transport for [npm/tink](https://github.com/npm/tink) [@achingbrain](https://github.com/achingbrain)
+
```
Comment (daviddias) [inline OKR/JS_CORE.md]: Suggestion: This objective should contain a KR on "js-ipfs can find (Peer Routing) any peer and dial to it (direct or relay) if this Peer is online."

Potentially adding a job that keeps trying to dial to many different nodes in the network, just to be sure that the connection always happens.

File: OKR/JS_CORE.md (hunk 2 above)
Comment (alanshaw) [reply OKR/JS_CORE.md]: Once we settle on a new CI this'll involve:

1. Figuring out a strategy to use e.g. solving questions like:
    * does publish a prerelease or a release?
    * how does CI know what version number to increment?
    * when does a release happen? push to master?
1. Configuring the CI to enact the release ensuring secrets are handled correctly
1. Testing the setup beforehand on a test module

File: OKR/JS_CORE.md (hunk 3 above)
Comment (alanshaw) [reply OKR/JS_CORE.md]: > Potentially adding a job that keeps trying to dial to many different nodes in the network, just to be sure that the connection always happens.

This is to mitigate against a failed connection due to network error or your laptop goes offline or something? I'm surprised this isn't already implemented!

**Comment by alanshaw** [comment]
> Is that something we could do some ground work for in Q1?

Absolutely. I think giving users feedback when IPFS "goes away" searching for something would be really valuable. Some suggestions for groundwork tasks that could usefully be done:

* A proposal on `interface-ipfs-core` for this feature
* Engage Go IPFS, and all libp2p teams to check they are on board with adding the feature
* Check it is possible (how would it work over HTTP?) and would not be too invasive
* Some dream code™️ for what the API would look like...on the JS side and I guess you could get someone from the Go team to imagineer how it could work in go

**Comment by alanshaw** [comment]
@achingbrain do you approve these OKRs?

**Comment by jbenet** [comment]
can't see retro doc-- not sure what perms should be (public or js-ipfs core dev team only)

**Comment by daviddias** [comment]
@jbenet the current default is WG only and invitees on share request.

File: OKR/JS_CORE.md [hunk 4]
```diff
@@ -2,6 +2,28 @@
 
 We frame our ongoing work using a process based on quarterly Objectives and Key Results (OKRs). Objectives reflect outcomes that are challenging, but realistic. Results are tangible and measurable.
 
+## 2019 Q1
+
+### A complete implementation of IPFS in JavaScript
+
+* p0 - Base32 encoded version 1 CIDs are the default for new content added to IPFS [@alanshaw](https://github.com/alanshaw)
+* p0 - IPNS has a streaming API [@vasco-santos](https://github.com/vasco-santos)
```
Comment (lidel) [inline OKR/JS_CORE.md]: :+1: as this would enable https://github.com/ipfs-shipyard/ipfs-companion/issues/643 with js-ipfs :)

**Comment by lidel** [comment]
>  3\. Use the machine or set or machines as the default gateway in a tool like companion (@lidel will this be possible? - perhaps in the beta channel?) to further test in real world scenarios.

If infra sets up a dedicated subdomain that will only route to JS-based gateways (eg. `js.gateway.ipfs.io` & `go.gateway.ipfs.io`), then we can use it as the default public gateway in ipfs-companion / ipfs-desktop (for all or  n% of users).

**Comment by momack2** [comment]
Hi JS Core WG! Your OKRs look really great and already updated with the [Q1 OKR sheet](https://docs.google.com/spreadsheets/d/1BtOfd7s9oYO5iKsIorCpsm4QuQoIsoZzSz7GItE-9ys/edit?ts=5c2f3d49#gid=274358435) (and they all have owners! 🙌) - can we merge this PR now and freeze OKRs for the quarter?

**Comment by daviddias** [comment]
@alanshaw this looks great! Thanks for getting the OKRs finalized.

Just one tiny tiddy little bit. Can you remove the OKRs from https://github.com/ipfs/team-mgmt/blob/master/OKR/JS_CORE.md and just point to the spreadsheet as a single source of truth?

**Comment by alanshaw** [comment]
sorry about that - https://github.com/ipfs/team-mgmt/pull/862

**Pull request merged**